sklearn/utils/testing.py

```diff
@@ -393,6 +403,177 @@ def assert_raise_message(exception, message, function, *args, **kwargs):
+                   " different results:\n\n%r\n\n and\n\n%r\n\n for :\n\n%r"
+                   "\n\nand :\n\n%r" % (method, res1, res2,
+                                        estimator1, estimator2))
+
+        assert same_model, msg
+
+
+def _not_same_model_method(method, X, estimator1, estimator2):
+    """Return True if both estimators return same result for given method"""
+    if hasattr(estimator1, method):
+        m = '%r\n\nhas %s, but\n\n%r\n\ndoes not' % (estimator1,
+                                                     method,
+                                                     estimator2)
+
+        if not hasattr(estimator2, method):
+            raise AttributeError(m)
+
+        # Check if the method(X) returns the same result for both models.
+        res1 = getattr(estimator1, method)(X)
+        res2 = getattr(estimator2, method)(X)
+        not_same_model = (res1.shape != res2.shape) or not np.allclose(res1,
+                                                                       res2)
+
+        # NOTE Do not assert here. This is to facilitate checking using any()
+        # of the results of multiple checks.
+        return not_same_model
+    return False  # if the method does not exist.
+
+
+def assert_same_model(X, estimator1, estimator2, msg=None):
+    """Helper function to check if models are same"""
+    _assert_same_model_method('predict', X, estimator1, estimator2, msg)
+    _assert_same_model_method('transform', X, estimator1, estimator2, msg)
+    _assert_same_model_method('decision_function',
+                              X, estimator1, estimator2, msg)
+    _assert_same_model_method('predict_proba',
+                              X, estimator1, estimator2, msg)
+
+    assert_fitted_attributes_equal(estimator1, estimator2)
+
+
+def assert_not_same_model(X, estimator1, estimator2, msg=None):
+    """Helper function to check if models are different"""
+    t1 = _not_same_model_method('predict', X, estimator1, estimator2)
+    t2 = _not_same_model_method('transform', X, estimator1, estimator2)
+    t3 = _not_same_model_method('decision_function', X, estimator1, estimator2)
+    t4 = _not_same_model_method('predict_proba', X, estimator1, estimator2)
+
+    if msg is None:
+        msg = ("Models are equal. Both models returned the same results for"
+               " all the methods.")
+
+    any_method_differed = any((t1, t2, t3, t4))
+
+    if not any_method_differed:
+        # If all methods return similar results for both ests or if those
+        # methods are not present
+        assert_fitted_attributes_not_equal(estimator1, estimator2)
+
+
+def _attributes_equal(estimator1, estimator2):
+    """Helper function to check if fitted model attributes are equal or not
+
+    Raises AssertionError if the attributes are not equal.
+    """
+
+    # A list of attributes which are known to be inconsistent.
+    skip_attributes = ('embedding_', 'weights_')
```

Comment:
In sklearn/utils/testing.py 
<https://github.com/scikit-learn/scikit-learn/pull/3907#discussion_r30912104>: 

> > +    any_method_differed = any((t1, t2, t3, t4))
> > +
> > +    if not any_method_differed:
> > +        # If all methods return similar results for both ests or if those
> > +        # methods are not present
> > +        assert_fitted_attributes_not_equal(estimator1, estimator2)
> > +
> > +
> > +def _attributes_equal(estimator1, estimator2):
> > +    """Helper function to check if fitted model attributes are equal or not
> > +
> > +    Raises AssertionError if the attributes are not equal.
> > +    """
> > +
> > +    # A list of attributes which are known to be inconsistent.
> > +    skip_attributes = ('embedding_', 'weights_')
>
> Spectral embedding is still non-deterministic due to the zero eigen 
> value problem which I am attempting to solve at #4299 
> <https://github.com/scikit-learn/scikit-learn/pull/4299> ;)
>
> I forgot which algorithm uses this |weights_| attribute ;) will 
> comment why I put the same in this list shortly :)
>
>
You should not skip it for all estimators. Hard-coding this in this 
place is really bad!